src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse42.cs

```diff
@@ -18,259 +18,386 @@ public static class Sse42
         /// <summary>
         /// int _mm_cmpistra (__m128i a, __m128i b, const int imm8)
         ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistrc (__m128i a, __m128i b, const int imm8)
-        ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistro (__m128i a, __m128i b, const int imm8)
-        ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistrs (__m128i a, __m128i b, const int imm8)
+        /// </summary>
+        public static bool CompareNotCAndNotZFlag(Vector128<sbyte> left, Vector128<sbyte> right, StringComparisonMode mode) => CompareNotCAndNotZFlag(left, right, mode);
```

Comment:
In case my original reply gets lost in the `outdated` section, I've copied it here:

I assume `C` == `Carry`, `Z` == `Zero`, `O` == `Overflow`, and `S` == `Sign`?  Can we spell out the words?  Or even better, if we could come up a more descriptive name of what the method does, that would be great.

Reading https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&text=_mm_cmpestr&expand=814,813,814,813, the only difference between `S` and `Z` that I can tell is which argument to check for a null character:

```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in a was null, and 0 otherwise. 
```
vs.
```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise. 
```

It seems the name should indicate/describe the differences, instead of using `SFlag` or `ZFlag`, and then forcing the caller to look up what `SFlag` means.

Reply:
> I assume C == Carry, Z == Zero, O == Overflow, and S == Sign? Can we spell out the words?

@eerhardt Right, and I am OK to spell them out if it matches the C# naming convention better.

> Or even better, if we could come up a more descriptive name of what the method does, that would be great.

I do not think so. "A more descriptive name" may be too long and not clear. For example, `NotCAndNotZFlag` looks even clearer than "absolute value of lb is larger than or equal to MaxSize and the resulting mask is equal to zero". Meanwhile, we already assume that the users of HW intrinsics have been familar/mastering with the C++ counterparts or ISA design. So, I still suggest to name these intrinsics with flags.

Reply:
I still think we can come up with better names.  It appears there are 5 cases we need to name:

| Flag   |      Description      |  Possible Names |
|----------|:-------------:|------:|
| S <br>&<br> Z |  Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if any character in a was null, and 0 otherwise. <br>&<br> Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise.  | CompareLeftContainsNull, CompareReturnLeftContainsNull, CompareReturnPartialLeft, CompareGetIsPartialLeft, CompareIsLeftPartial <br>(Same for `Right`) |
| O | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns bit 0 of the resulting bit mask.  |    CompareReturnFirstBit, CompareGetFirstBit |
| C | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if the resulting mask was non-zero, and 0 otherwise.  |    CompareIsResultNonZero, CompareHasNonZeroResult, CompareNonZeroMask, CompareHasResult  |
| A | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if b did not contain a null character and the resulting mask was zero, and 0 otherwise.   |      |

`A` is definitely hard to describe succinctly, since it is the combination of two other ones.  So coming up with a good name for it will probably be based on the two underlying names.

-------
**Note**: The [IntrinsicsGuide's description](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&expand=883,2543,936,814&techs=SSE4_2&text=_mm_cmp) for the explicit length `S` & `Z` overloads (`_mm_cmpestrs` & `_mm_cmpestrz`) appears to be incorrect.
`Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise. `
But then the Operation says:
```
size := (imm8[0] ? 16 : 8) // 8 or 16-bit characters
UpperBound := (128 / size) - 1

dst := (lb <= UpperBound)
```
Where the Operation appears to match what is in the underlying instruction manuals.

-------

In light of the implicit vs explicit differences, to name `S` & `Z`, that's why I like something like `Is Partial` because it describes at a higher-level what is being checked.  ex. `File.Exists` instead of `File.StatReturnsZero`.

Reply:
@eerhardt Thank you so much for the investigation. I read the manual again and I agree that "a more descriptive name" is very necessary because FLAGS is used in a non-standard manner with these SSE4.2 string instructions.
According to the above table you provided, there is a potential name matrix:

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareNonZeroResultMask |
| CompareZFlag | CompareRightContainsNull |
| CompareSFlag | CompareLeftContainsNull |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareZeroResultMaskAndRightNotContainsNull|

Does it look good to you?

cc @CarolEidt @tannergooding

Reply:
The only issue I see if the difference between implicit and explicit length overloads. If we called the explicit length overloads `ContainsNull` that would be incorrect. It instead returns if the `rightLength` is less than the maximum (8 or 16). So we should use a different name for the explicit length overloads.

I'd like to keep them using the same name, if a decent name comes up. But if one isn't available, I'm fine with the implicit and explicit length overloads being named different (`RightContainsNull` vs. `RightLessThanMax`/`RightIsPartial`/etc.)

Reply:
I also don't like `ZeroResultMask` as the name here... `CompareEqual` or `IsMatch` might be better?

Reply:
How about `HasMatch` or `ContainsMatch`? Since depending on the string comparison mode, the whole thing may not match.

Reply:
> I also don't like ZeroResultMask as the name here... CompareEqual or IsMatch might be better?

Zero result masks do not always mean equal or not, which may be controlled by `StringComparisonMode` (e.g., `EqualAny`, `Ranges`, `XXXPolarity`, etc).

Reply:
`HasResult`/`ContainsResult`?

Reply:
I think we could make it more obvious by calling the `Polarity` flags something like `NegateResult` instead.

Reading `if (ContainsMatch(op1, op2, StringComparisonMode.NegateResult))` seems much better (both readable and understandable) than `if (CompareZeroResultMask(op1, op2, StringComparisonMode.NegatiePolarity))`

Reply:
> We could do EndOfString, since that applies to both Explicit (where end of string is length determined) and Implicit (where end of string is determined by the null character)?

I think this is a good idea and I also perfer to use the same name for Explicit and Implicit.
How about `Left/RightTerminated`?

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareNonZeroResultMask |
| CompareZFlag | CompareRightTerminated |
| CompareSFlag | CompareLeftTerminated |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareZeroResultMaskAndRightNotTerminated|

Reply:
> I think we could make it more obvious by calling the Polarity flags something like NegateResult instead.

Good point! We can rename these two polarity flags both:
```csharp
        /// <summary>
        /// _SIDD_NEGATIVE_POLARITY
        /// </summary>
        NegativeResult = 0x10,

        /// <summary>
        /// _SIDD_MASKED_NEGATIVE_POLARITY
        /// </summary>
        NegativeUsefulResult = 0x30,
```

Reply:
With the above renamed StringComparisonMode, we seems to have the new name matrix:

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareHasMatch |
| CompareZFlag | CompareRightTerminated |
| CompareSFlag | CompareLeftTerminated |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareNoMatchAndRightNotTerminated|